lighthouse/utils/runner.py

```diff
@@ -0,0 +1,166 @@
+    payload_module = workload.lower_payload(schedule_parameters=schedule_parameters)
+    # get execution engine
+    engine = get_engine(payload_module, shared_libs=workload.shared_libs())
+
+    with workload.allocate_inputs(execution_engine=engine) as inputs:
+        # prepare function arguments
+        packed_args = memrefs_to_packed_args(inputs)
+
+        # handle to payload function
+        payload_func = engine.lookup(workload.payload_function_name)
+
+        # call function
+        payload_func(packed_args)
+
+        if check_correctness:
+            success = workload.check_correctness(
+                execution_engine=engine, verbose=verbose
+            )
+            if not success:
+                raise ValueError("Benchmark verification failed.")
+
+
+def emit_benchmark_function(
+    payload_module: ir.Module,
+    workload: Workload,
+    nruns: int,
+    nwarmup: int,
+):
+    """
+    Emit a benchmark function that calls payload function and times it.
+
+    Every function call is timed separately. Returns the times (seconds) in a
+    memref.
+    """
+    # find original payload function
+    payload_func = None
+    for op in payload_module.operation.regions[0].blocks[0]:
+        if (
+            isinstance(op, func.FuncOp)
+            and op.name.value == workload.payload_function_name
+        ):
+            payload_func = op
+            break
+    assert payload_func is not None, "Could not find payload function"
+    payload_arguments = payload_func.type.inputs
+
+    # emit benchmark function that calls payload and times it
+    with ir.InsertionPoint(payload_module.body):
+        # define rtclock function
+        f64_t = ir.F64Type.get()
+        func.FuncOp("rtclock", ((), (f64_t,)), visibility="private")
+        # emit benchmark function
+        time_memref_t = ir.MemRefType.get((nruns,), f64_t)
+        args = payload_arguments + [time_memref_t]
+
+        @func.func(*args)
+        def benchmark(*args):
+            index_t = ir.IndexType.get()
+            zero = arith.constant(index_t, 0)
+            one = arith.constant(index_t, 1)
+            nwarmup_cst = arith.constant(index_t, nwarmup)
+            for i in scf.for_(zero, nwarmup_cst, one):
+                # FIXME(upstream): func.call is broken for this use case?
+                func.CallOp(payload_func, list(args[: len(payload_arguments)]))
+                scf.yield_(())
+            nruns_cst = arith.constant(index_t, nruns)
+            for i in scf.for_(zero, nruns_cst, one):
+                tic = func.call((f64_t,), "rtclock", ())
+                func.CallOp(payload_func, list(args[: len(payload_arguments)]))
+                toc = func.call((f64_t,), "rtclock", ())
+                time = arith.subf(toc, tic)
+                memref.store(time, args[-1], [i])
+                scf.yield_(())
+
+        benchmark.func_op.attributes["llvm.emit_c_interface"] = ir.UnitAttr.get()
+
+
+def benchmark(
```

Comment:
This feels like it is implementing a little "protocol" for how to interact with `Workload` objects. Given its close tie to `Workload`, we could add it as a method on `Workload`. We could even mark that method `@final`: https://peps.python.org/pep-0591/#id1

Same goes for `execute`, I think.

(Sorry if I am repeating myself from a couple weeks back.)

Reply:
For me personally, such protocol-implementing functions also help me understand the different parts of an interface, e.g. which purpose(s) they serve. Having them close to the definition of the interface could help people (like me).